azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -128,40 +122,64 @@ def update_host_plugin_headers(wire_client):
         """
-        timestamp = create_timestamp()
-        xml_text, xml_doc, incarnation = GoalState._fetch_goal_state(self._wire_client)
-
-        if force_update or self._incarnation != incarnation:
-            # If we are fetching a new goal state
-            self._history = GoalStateHistory(timestamp, incarnation)
-            self._initialize_basic_properties(xml_doc)
-            self._fetch_extended_goal_state(xml_text, xml_doc, force_vm_settings_update=force_update)
-        else:
-            # else ensure the extensions are using the latest vm_settings
-            timestamp = create_timestamp()
-            vm_settings, vm_settings_updated = self._fetch_vm_settings(force_update=force_update)
+        #
+        # Fetch the goal state from both the HGAP and the WireServer
+        #
+        timestamp = timeutil.create_timestamp()
+
+        incarnation, xml_text, xml_doc = GoalState._fetch_goal_state(self._wire_client)
+        goal_state_updated = incarnation != self._incarnation
+        if goal_state_updated:
+            logger.info('Fetched new goal state from the WireServer [incarnation {0}]', incarnation)
+
+        vm_settings, vm_settings_updated = GoalState._fetch_vm_settings(self._wire_client)
+        if vm_settings_updated:
+            logger.info("Fetched new vmSettings [HostGAPlugin correlation ID: {0} eTag: {1} source: {2}]", vm_settings.hostga_plugin_correlation_id, vm_settings.etag, vm_settings.source)
+        # Ignore the vmSettings if their source is Fabric (processing a Fabric goal state may require the tenant certificate and the vmSettings don't include it.)
+        if vm_settings is not None and vm_settings.source == GoalStateSource.Fabric:
             if vm_settings_updated:
-                logger.info("Fetched new vmSettings [HostGAPlugin correlation ID: {0} eTag: {1} source: {2}]", vm_settings.hostga_plugin_correlation_id, vm_settings.etag, vm_settings.source)
-                self._history = GoalStateHistory(timestamp, vm_settings.etag)
-                self._extensions_goal_state = vm_settings
-                self._history.save_vm_settings(vm_settings.get_redacted_text())
+                logger.info("The vmSettings originated via Fabric; will ignore them.")
+            vm_settings, vm_settings_updated = None, False
+
+        # If neither goal state has changed we are done with the update
+        if not goal_state_updated and not vm_settings_updated:
+            return
+
+        # Start a new history subdirectory and capture the updated goal state
+        tag = "{0}".format(incarnation) if vm_settings is None else "{0}-{1}".format(incarnation, vm_settings.etag)
+        self._history = GoalStateHistory(timestamp, tag)
+        if goal_state_updated:
+            self._history.save_goal_state(xml_text)
+        if vm_settings_updated:
+            self._history.save_vm_settings(vm_settings.get_redacted_text())
```

Comment:
We used to keep both `extension_config` and `vm_settings` on fabric goal state. Just to confirm with this change, we longer have vm_settings in history folder even `vm_settings` updated?

Reply:
Not sure I understand the question.

We were using the HostGAPlugin both for Fabric and FastTrack goal states, and we were always saving the vmSettings and ExtensionsConfig (the latter just for debugging purposes).

Since now we ignore Fabric goal states from the HostGAPlugin, I do not save those to the history folder. FastTrack goal states should still be saved.

Reply:
Sample from a DCR run

```
./2022-03-31T22_26_21.054052_4:
    -rwxrwxrwx 1 nam nam 2800 Mar 31 22:26 ExtensionsConfig.xml
    -rwxrwxrwx 1 nam nam 2099 Mar 31 22:26 GoalState.xml
    -rwxrwxrwx 1 nam nam 1441 Mar 31 22:26 HostingEnvironmentConfig.xml
    -rwxrwxrwx 1 nam nam 7406 Mar 31 22:26 Microsoft.Azure.Extensions.CustomScript.manifest.xml
    -rwxrwxrwx 1 nam nam  899 Mar 31 22:26 SharedConfig.xml
    -rwxrwxrwx 1 nam nam 1020 Mar 31 22:26 waagent_status.json

./2022-03-31T22_27_01.331737_5:
    -rwxrwxrwx 1 nam nam  3187 Mar 31 15:27 ExtensionsConfig.xml
    -rwxrwxrwx 1 nam nam  2099 Mar 31 15:27 GoalState.xml
    -rwxrwxrwx 1 nam nam  1721 Mar 31 15:27 HostingEnvironmentConfig.xml
    -rwxrwxrwx 1 nam nam 13020 Mar 31 15:27 Microsoft.OSTCExtensions.VMAccessForLinux.manifest.xml
    -rwxrwxrwx 1 nam nam   899 Mar 31 15:27 SharedConfig.xml
    -rwxrwxrwx 1 nam nam  1632 Mar 31 15:27 waagent_status.json

./2022-03-31T22_27_47.722244_5-11145155257095352179:
    -rwxrwxrwx 1 nam nam 13020 Mar 31 15:27 Microsoft.OSTCExtensions.VMAccessForLinux.manifest.xml
    -rwxrwxrwx 1 nam nam  2299 Mar 31 15:27 VmSettings.json
    -rwxrwxrwx 1 nam nam  1018 Mar 31 15:28 waagent_status.json

./2022-03-31T22_28_58.078953_5-1812135654488544646:
    -rwxrwxrwx 1 nam nam 1301 Mar 31 15:28 Microsoft.CPlat.Core.RunCommandHandlerLinux.manifest.xml
    -rwxrwxrwx 1 nam nam 2527 Mar 31 15:28 VmSettings.json
    -rwxrwxrwx 1 nam nam 1845 Mar 31 15:29 waagent_status.json

```

Reply:
Sure, what name would you suggest? we need to keep short and should be easy to correlate with the agent's log (that is why we use the timestamp-incarnation-etag combo). Also, should not be too different so that we can capture it easily with the regexes in archive.py that also match the current name (otherwise the archiving code would get more complex)

Reply:
I was thinking something like this `2022-03-31T22_27_47.722244_incar_5-etag_11145155257095352179`. I understand it's too long, but this makes things clearer. I don't know how difficult it's to change pattern in archive.py. Anyway, it's just my suggestion but leave this to make you a decision.

Reply:
ok, i'll leave it as is. at some point i also considered adding "incarnation" and "etag", but they make the name longer and once i started debugging issues i realized the extra text was not adding anything useful,  so i removed it.

Reply:
Or do you mean that the incarnation is so short it gets confused with the timestamp? I could add another "_" and remove the milliseconds from the timestamp: 


`
2022-03-31T16:11:32__2-7855907865217799575.zip
`
I just realized my example has underscores in the timestamp. That example was coming from a ZIP expanded on Windows. On the actual VM the timestamp would actually have colons, as in this last example. I'm considering replacing those with dashes, because the shell's tab completion doesn't work well with colons. So it would be something like this
`

`
2022-03-31T16-11-32__2-7855907865217799575.zip
`

I would do those changes on a separate PR, though, since the archive code and corresponding tests would need some rewrtiting.

Reply:
```
Or do you mean that the incarnation is so short it gets confused with the timestamp? I could add another "_" and remove the milliseconds from the timestamp:

2022-03-31T16:11:32__2-7855907865217799575.zip
```

yeah, this looks better and clearer. Thanks for the change.